README


This archive contains the distribution Throwable,
version 0.200012:

  a role for classes that can be thrown

This software is copyright (c) 2014 by Ricardo SIGNES.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


This README file was generated by Dist::Zilla::Plugin::Readme v5.020.



Changes
Revision history for Throwable

0.200012  2014-09-19 21:59:28-04:00 America/New_York
        - avoid memory leaks and infinite recursion when stringifying
          (thanks, Dagfinn Ilmari Mannsåker!)

0.200011  2014-04-02 20:49:49-04:00 America/New_York
        - eliminate prereq on Class::Load (thanks, Toby Inkster!)
        - eliminate preqre on MooX::Types::MooseLike::Base (Toby Inkster)

0.200010  2014-03-08 22:39:25-05:00 America/New_York
        - eliminate a memory leak in StackTrace::Auto (thanks, Graham Knop!)

0.200009  2013-10-21 22:05:53 America/New_York
        - avoid boolifying objects that might be held in the out of band
          storage
        - update bugtracker and repo metadata
        - more changes coming soon

0.200008  2013-04-30 12:50:19 America/New_York
          avoid a Moose-using test in the suite of tests run by users (thanks,
          Dave Rolsky)

0.200007  2013-04-09 19:20:02 America/New_York
          allow previous_exception to be passed to the constructor

0.200006  2013-04-09 14:58:07 America/New_York
          lose old $@ in previous_exception less often under 5.17.10 (thanks,
          David Wheeler)

0.200005  2013-04-05 15:04:52 America/New_York
          Plugged memory leak (thaljef)

0.200004  2013-02-21 16:50:05 America/New_York
          update Throwable::Error to use around modifier instead of calling
          SUPER:: (thanks, Justin Hunter)

0.200003  2012-10-18 08:52:20 America/New_York
          no code changes; just no longer trial

0.200002  2012-10-09 08:56:21 America/New_York
          require Moo 1.000001

0.200001  2012-09-22 18:10:29 America/New_York
          update a prerequisite (thanks, FREW)

0.200000  2012-09-20 14:12:34 America/New_York
          port to Moo (FREW, MSTROUT, ILMARI)

0.102080  2010-07-27 12:09:21 America/New_York
          refactor stack trace autocreation to StackTrace::Auto role
          make Throwable::Error immutable

0.101110  2010-04-21 20:59:59 America/New_York
          avoid doing: length undef (thanks, Yuval Kogman)

0.100090  2010-01-10
          $error->throw now rethrows existing Errors
          Error->throw("str") now uses the string as message attr

0.092610  2009-09-18
          fix to work with latest Class::MOP (thanks, Paul Mooney)

0.092000  2009-07-18
          fix required version of Moose to avoid failures

0.091900  2009-07-09
          rely on the stable release of Devel::StackTrace that does what we
          need

0.091780  2009-06-26
          first testing release


Makefile.PL

# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.020.
use strict;
use warnings;



use ExtUtils::MakeMaker ;



my %WriteMakefileArgs = (
  "ABSTRACT" => "a role for classes that can be thrown",
  "AUTHOR" => "Ricardo SIGNES <rjbs\@cpan.org>, Florian Ragwitz <rafl\@debian.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "Throwable",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "Throwable",
  "PREREQ_PM" => {
    "Carp" => 0,
    "Devel::StackTrace" => "1.32",
    "Module::Runtime" => "0.002",
    "Moo" => "1.000001",
    "Moo::Role" => 0,
    "Scalar::Util" => 0,
    "Sub::Quote" => 0,
    "overload" => 0
  },
  "TEST_REQUIRES" => {
    "Devel::StackTrace" => "1.32",
    "Test::More" => "0.96",
    "base" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "VERSION" => "0.200012",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Carp" => 0,
  "Devel::StackTrace" => "1.32",
  "Module::Runtime" => "0.002",
  "Moo" => "1.000001",
  "Moo::Role" => 0,
  "Scalar::Util" => 0,
  "Sub::Quote" => 0,
  "Test::More" => "0.96",
  "base" => 0,
  "overload" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);





lib/Throwable.pm
package Throwable;
# ABSTRACT: a role for classes that can be thrown
$Throwable::VERSION = '0.200010';
use Moo::Role;
use Sub::Quote ();
use Scalar::Util ();
use Carp ();

# =head1 SYNOPSIS
#
#   package Redirect;
#   use Moose;
#   with 'Throwable';
#
#   has url => (is => 'ro');
#
# ...then later...
#
#   Redirect->throw({ url => $url });
#
# =head1 DESCRIPTION
#
# Throwable is a role for classes that are meant to be thrown as exceptions to
# standard program flow.  It is very simple and does only two things: saves any
# previous value for C<$@> and calls C<die $self>.
#
# Throwable is implemented with L<Moo>, so you can stick to Moo or use L<Moose>
# as you prefer.
#
# =attr previous_exception
#
# This attribute is created automatically, and stores the value of C<$@> when the
# Throwable object is created.  This is done on a I<best effort basis>.  C<$@> is
# subject to lots of spooky action-at-a-distance.  For now, there are clearly
# ways that the previous exception could be lost.
#
# =cut

our %_HORRIBLE_HACK;

has 'previous_exception' => (
  is       => 'ro',
  default  => Sub::Quote::quote_sub(q<
    if (defined $Throwable::_HORRIBLE_HACK{ERROR}) {
      $Throwable::_HORRIBLE_HACK{ERROR}
    } elsif (defined $@ and (ref $@ or length $@)) {
      $@;
    } else {
      undef;
    }
  >),
);

# =method throw
#
#   Something::Throwable->throw({ attr => $value });
#
# This method will call new, passing all arguments along to new, and will then
# use the created object as the only argument to C<die>.
#
# If called on an object that does Throwable, the object will be rethrown.
#
# =cut

sub throw {
  my ($inv) = shift;

  if (Scalar::Util::blessed($inv)) {
    Carp::confess "throw called on Throwable object with arguments" if @_;
    die $inv;
  }

  local $_HORRIBLE_HACK{ERROR} = $@;
  my $throwable = $inv->new(@_);
  die $throwable;
}

no Moo::Role;
1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Throwable - a role for classes that can be thrown

=head1 VERSION

version 0.200010

=head1 SYNOPSIS

  package Redirect;
  use Moose;
  with 'Throwable';

  has url => (is => 'ro');

...then later...

  Redirect->throw({ url => $url });

=head1 DESCRIPTION

Throwable is a role for classes that are meant to be thrown as exceptions to
standard program flow.  It is very simple and does only two things: saves any
previous value for C<$@> and calls C<die $self>.

Throwable is implemented with L<Moo>, so you can stick to Moo or use L<Moose>
as you prefer.

=head1 ATTRIBUTES

=head2 previous_exception

This attribute is created automatically, and stores the value of C<$@> when the
Throwable object is created.  This is done on a I<best effort basis>.  C<$@> is
subject to lots of spooky action-at-a-distance.  For now, there are clearly
ways that the previous exception could be lost.

=head1 METHODS

=head2 throw

  Something::Throwable->throw({ attr => $value });

This method will call new, passing all arguments along to new, and will then
use the created object as the only argument to C<die>.

If called on an object that does Throwable, the object will be rethrown.

=head1 AUTHORS

=over 4

=item *

Ricardo SIGNES <rjbs@cpan.org>

=item *

Florian Ragwitz <rafl@debian.org>

=back

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Ricardo SIGNES.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/StackTrace/Auto.pm
package StackTrace::Auto;
# ABSTRACT: a role for generating stack traces during instantiation
$StackTrace::Auto::VERSION = '0.200010';
use Moo::Role;
use Sub::Quote ();
use MooX::Types::MooseLike::Base qw(ArrayRef);
use Class::Load 0.20 ();
use Scalar::Util ();

# =head1 SYNOPSIS
#
# First, include StackTrace::Auto in a Moose class...
#
#   package Some::Class;
#   use Moose;
#   with 'StackTrace::Auto';
#
# ...then create an object of that class...
#
#   my $obj = Some::Class->new;
#
# ...and now you have a stack trace for the object's creation.
#
#   print $obj->stack_trace->as_string;
#
# =attr stack_trace
#
# This attribute will contain an object representing the stack at the point when
# the error was generated and thrown.  It must be an object performing the
# C<as_string> method.
#
# =attr stack_trace_class
#
# This attribute may be provided to use an alternate class for stack traces.  The
# default is L<Devel::StackTrace|Devel::StackTrace>.
#
# In general, you will not need to think about this attribute.
#
# =cut

has stack_trace => (
  is       => 'ro',
  isa      => Sub::Quote::quote_sub(q{
    require Scalar::Util;
    die "stack_trace must be have an 'as_string' method!" unless
       Scalar::Util::blessed($_[0]) && $_[0]->can('as_string')
  }),
  default  => Sub::Quote::quote_sub(q{
    $_[0]->stack_trace_class->new(
      @{ $_[0]->stack_trace_args },
    );
  }),
  init_arg => undef,
);

has stack_trace_class => (
  is      => 'ro',
  isa     => Sub::Quote::quote_sub(q{
      die "stack_trace_class must be a loaded class"
          unless Class::Load::is_class_loaded($_[0]);
  }),
  coerce  => Sub::Quote::quote_sub(q{
    Class::Load::load_class($_[0]);
  }),
  lazy    => 1,
  builder => '_build_stack_trace_class',
);

# =attr stack_trace_args
#
# This attribute is an arrayref of arguments to pass when building the stack
# trace.  In general, you will not need to think about it.
#
# =cut

has stack_trace_args => (
  is      => 'ro',
  isa     => ArrayRef,
  lazy    => 1,
  builder => '_build_stack_trace_args',
);

sub _build_stack_trace_class {
  return 'Devel::StackTrace';
}

sub _build_stack_trace_args {
  my ($self) = @_;

  Scalar::Util::weaken($self);  # Prevent memory leak

  my $found_mark = 0;
  return [
    frame_filter => sub {
      my ($raw) = @_;
      my $sub = $raw->{caller}->[3];
      (my $package = $sub) =~ s/::\w+\z//;
      if ($found_mark == 2) {
          return 1;
      }
      elsif ($found_mark == 1) {
        return 0 if $sub =~ /::new$/ && $self->isa($package);
        $found_mark++;
        return 1;
      } else {
        $found_mark++ if $sub =~ /::new$/ && $self->isa($package);
        return 0;
      }
    },
  ];
}

no Moo::Role;
1;

__END__

=pod

=encoding UTF-8

=head1 NAME

StackTrace::Auto - a role for generating stack traces during instantiation

=head1 VERSION

version 0.200010

=head1 SYNOPSIS

First, include StackTrace::Auto in a Moose class...

  package Some::Class;
  use Moose;
  with 'StackTrace::Auto';

...then create an object of that class...

  my $obj = Some::Class->new;

...and now you have a stack trace for the object's creation.

  print $obj->stack_trace->as_string;

=head1 ATTRIBUTES

=head2 stack_trace

This attribute will contain an object representing the stack at the point when
the error was generated and thrown.  It must be an object performing the
C<as_string> method.

=head2 stack_trace_class

This attribute may be provided to use an alternate class for stack traces.  The
default is L<Devel::StackTrace|Devel::StackTrace>.

In general, you will not need to think about this attribute.

=head2 stack_trace_args

This attribute is an arrayref of arguments to pass when building the stack
trace.  In general, you will not need to think about it.

=head1 AUTHORS

=over 4

=item *

Ricardo SIGNES <rjbs@cpan.org>

=item *

Florian Ragwitz <rafl@debian.org>

=back

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Ricardo SIGNES.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/Throwable/Error.pm
package Throwable::Error;
# ABSTRACT: an easy-to-use class for error objects
$Throwable::Error::VERSION = '0.200010';
use Moo 1.000001;
use MooX::Types::MooseLike::Base qw(Str);
with 'Throwable', 'StackTrace::Auto';

# =head1 SYNOPSIS
#
#   package MyApp::Error;
#   use Moose;
#   extends 'Throwable::Error';
#
#   has execution_phase => (
#     is  => 'ro',
#     isa => 'MyApp::Phase',
#     default => 'startup',
#   );
#
# ...and in your app...
#
#   MyApp::Error->throw("all communications offline");
#
#   # or...
#
#   MyApp::Error->throw({
#     message         => "all communications offline",
#     execution_phase => 'shutdown',
#   });
#
# =head1 DESCRIPTION
#
# Throwable::Error is a base class for exceptions that will be thrown to signal
# errors and abort normal program flow.  Throwable::Error is an alternative to
# L<Exception::Class|Exception::Class>, the features of which are largely
# provided by the Moose object system atop which Throwable::Error is built.
#
# Throwable::Error performs the L<Throwable|Throwable> and L<StackTrace::Auto>
# roles.  That means you can call C<throw> on it to create and throw an error
# object in one call, and that every error object will have a stack trace for its
# creation.
#
# =cut

use overload
  q{""}    => 'as_string',
  fallback => 1;

# =attr message
#
# This attribute must be defined and must contain a string describing the error
# condition.  This string will be printed at the top of the stack trace when the
# error is stringified.
#
# =cut

has message => (
  is       => 'ro',
  isa      => Str,
  required => 1,
);

# =attr stack_trace
#
# This attribute, provided by L<StackTrace::Auto>, will contain a stack trace
# object guaranteed to respond to the C<as_string> method.  For more information
# about the stack trace and associated behavior, consult the L<StackTrace::Auto>
# docs.
#
# =method as_string
#
# This method will provide a string representing the error, containing the
# error's message followed by the its stack trace.
#
# =cut

sub as_string {
  my ($self) = @_;

  my $str = $self->message;
  $str .= "\n\n" . $self->stack_trace->as_string;

  return $str;
}

around BUILDARGS => sub {
  my $orig = shift;
  my $self = shift;

  return {} unless @_;
  return {} if @_ == 1 and ! defined $_[0];

  if (@_ == 1 and (!ref $_[0]) and defined $_[0] and length $_[0]) {
    return { message => $_[0] };
  }

  return $self->$orig(@_);
};

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Throwable::Error - an easy-to-use class for error objects

=head1 VERSION

version 0.200010

=head1 SYNOPSIS

  package MyApp::Error;
  use Moose;
  extends 'Throwable::Error';

  has execution_phase => (
    is  => 'ro',
    isa => 'MyApp::Phase',
    default => 'startup',
  );

...and in your app...

  MyApp::Error->throw("all communications offline");

  # or...

  MyApp::Error->throw({
    message         => "all communications offline",
    execution_phase => 'shutdown',
  });

=head1 DESCRIPTION

Throwable::Error is a base class for exceptions that will be thrown to signal
errors and abort normal program flow.  Throwable::Error is an alternative to
L<Exception::Class|Exception::Class>, the features of which are largely
provided by the Moose object system atop which Throwable::Error is built.

Throwable::Error performs the L<Throwable|Throwable> and L<StackTrace::Auto>
roles.  That means you can call C<throw> on it to create and throw an error
object in one call, and that every error object will have a stack trace for its
creation.

=head1 ATTRIBUTES

=head2 message

This attribute must be defined and must contain a string describing the error
condition.  This string will be printed at the top of the stack trace when the
error is stringified.

=head2 stack_trace

This attribute, provided by L<StackTrace::Auto>, will contain a stack trace
object guaranteed to respond to the C<as_string> method.  For more information
about the stack trace and associated behavior, consult the L<StackTrace::Auto>
docs.

=head1 METHODS

=head2 as_string

This method will provide a string representing the error, containing the
error's message followed by the its stack trace.

=head1 AUTHORS

=over 4

=item *

Ricardo SIGNES <rjbs@cpan.org>

=item *

Florian Ragwitz <rafl@debian.org>

=back

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Ricardo SIGNES.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut
